server/src/main/java/org/elasticsearch/action/update/TransportUpdateAction.java

```diff
@@ -333,4 +343,58 @@ private void handleUpdateFailureWithRetry(
         }
         listener.onFailure(cause instanceof Exception ? (Exception) cause : new NotSerializableExceptionWrapper(cause));
     }
+
+    private static UpdateHelper.Result deleteInferenceResults(
+        UpdateRequest updateRequest,
+        UpdateHelper.Result result,
+        IndexMetadata indexMetadata,
+        MappingLookup mappingLookup
+    ) {
+        if (result.getResponseResult() != DocWriteResponse.Result.UPDATED) {
+            return result;
+        }
+
+        Map<String, InferenceFieldMetadata> inferenceFields = indexMetadata.getInferenceFields();
+        if (inferenceFields.isEmpty()) {
+            return result;
+        }
+
+        Map<String, Object> updateRequestSource = updateRequest.doc().sourceAsMap();
+        Map<String, Object> updatedSource = result.updatedSourceAsMap();
+        boolean updatedSourceModified = false;
+        for (var entry : inferenceFields.entrySet()) {
+            String inferenceFieldName = entry.getKey();
+            Mapper mapper = mappingLookup.getMapper(inferenceFieldName);
+
+            if (mapper instanceof InferenceFieldMapper inferenceFieldMapper) {
+                String[] sourceFields = entry.getValue().getSourceFields();
+                for (String sourceField : sourceFields) {
```

Comment:
Do we need to go through all the source fields? It looks like we only need to change the inference field so this loop is redundant.

Reply:
We need to check if any of the source fields were changed by the update request

Reply:
ah ok got it, so we do keep the inference results if none of the source field are present in the update, nice!